### PR TITLE
Uninstall functionality

### PR DESCRIPTION
**Pull request by mikusp**

### CODE REVIEW

File: luna-manager/src/Luna/Manager/Command/Uninstall.hs [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+createdByLunaStudio = ("LunaStudio" `Text.isInfixOf`) . Shelly.toTextIgnore . last . splitDirectories
+
+uninstallElectronCaches :: MonadUninstall m => m ()
+uninstallElectronCaches = when (currentHost /= Darwin) $ do
+    baseDir <- Text.pack <$> case currentHost of
+        Linux   -> return "~/.config"
+        Windows -> liftIO $ Dir.getAppUserDataDirectory ""
+        _       -> error "uninstallElectronCaches: unsupported system detected"
+    dirs <- Shelly.ls $ Shelly.fromText baseDir
+    -- NOTE[MM]: splitting a directory and taking last is a bit more involved
```
Comment (kustosz) [inline luna-manager/src/Luna/Manager/Command/Uninstall.hs]: I'm absolutely stunned by the fact you have considered this! 😍

File: luna-manager/src/Luna/Manager/Command/Uninstall.hs [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+default(Text.Text)
+
+type MonadUninstall m = (MonadIO m, MonadSh m, MonadShControl m, MonadCatch m, MonadGetter Install.InstallConfig m, MonadGetter Options m, MonadGetter EnvConfig m)
+
+uninstallServices :: MonadUninstall m => Install.InstallConfig -> m ()
+uninstallServices opts = case currentHost of
+    Windows -> do
+        let currentServices = (opts ^. Install.defaultBinPathGuiApp)
+                          </> Shelly.fromText "LunaStudio"
```
Comment (sylwiabr) [inline luna-manager/src/Luna/Manager/Command/Uninstall.hs]: "LunaStudio" should be an argument like appName to check which app you want delete, also services will be deleted just for "gui application" so if we delete luna by itself there is no services to uninstall

File: luna-manager/src/Luna/Manager/Command/Uninstall.hs [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+                          </> Shelly.fromText "current"
```
Comment (sylwiabr) [inline luna-manager/src/Luna/Manager/Command/Uninstall.hs]: this is `selectedVersionPath ` from `InstallConfig`

File: luna-manager/src/Luna/Manager/Command/Uninstall.hs [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+    confPath <- expand $ opts ^. Install.defaultConfPath
+    Logger.log $ "Removing local config from " <> Shelly.toTextIgnore confPath
+    Shelly.rm_rf confPath `Exception.catchAny` (\(e::SomeException) ->
+        Logger.warning $ "Removing local config from " <> Shelly.toTextIgnore confPath <> " failed "
+            <> "because of " <> convert (displayException e) <> ". Continuing...")
+
+createdByLunaStudio :: FilePath -> Bool
+createdByLunaStudio = ("LunaStudio" `Text.isInfixOf`) . Shelly.toTextIgnore . last . splitDirectories
+
+uninstallElectronCaches :: MonadUninstall m => m ()
```
Comment (sylwiabr) [inline luna-manager/src/Luna/Manager/Command/Uninstall.hs]: what about Darwin?

File: luna-manager/src/Luna/Manager/Command/Uninstall.hs [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+uninstallElectronCaches = when (currentHost /= Darwin) $ do
+    baseDir <- Text.pack <$> case currentHost of
+        Linux   -> return "~/.config"
```
Comment (sylwiabr) [inline luna-manager/src/Luna/Manager/Command/Uninstall.hs]: i think this should be taken from `System.Directory` as `getXdgDirectory` on all system

**Comment by sylwiabr** [comment]
If we want to install completely everything in my opinion the way to do is to download config check available applications and based on this try to remove all of them.

File: luna-manager/src/Luna/Manager/Command/Uninstall.hs (hunk 5 above)
Comment (mikusp) [reply luna-manager/src/Luna/Manager/Command/Uninstall.hs]: That's right. As I checked now, Electron respects $XDG_CONFIG_HOME

**Comment by mikusp** [comment]
I forgot that currently it cannot remove directories longer than MAX_PATH on Windows

File: luna-manager/src/Luna/Manager/Command/Uninstall.hs [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+-- NOTE[MM]: splitting a directory and taking last is a bit more involved
+--           than running Text.isInfixOf directly, but is more correct
+--           if e.g. username is LunaStudio - in that case, every path
+--           would be deleted
+createdByLunaStudio :: FilePath -> Bool
+createdByLunaStudio = (lunaStudio `Text.isInfixOf`) . Shelly.toTextIgnore . last . splitDirectories
+
+uninstallElectronCaches :: MonadUninstall m => m ()
+uninstallElectronCaches = when (currentHost /= Darwin) $ do
+    baseDir <- case currentHost of
```
[the remark this answers is not in the document]
Comment (mikusp) [reply luna-manager/src/Luna/Manager/Command/Uninstall.hs]: Not sure about that, cause I can't think of a good name for this directory and I don't think this function would be usable anywhere else. It's an electron quirk that it places its cache in directories that have completely different purpose

File: luna-manager/src/Luna/Manager/Command/Uninstall.hs [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+    Shelly.rm_rf confPath `Exception.catchAny` (\(e::SomeException) ->
+        Logger.warning $ "Removing local config from " <> Shelly.toTextIgnore confPath <> " failed "
+            <> "because of " <> convert (displayException e) <> ". Continuing...")
+
+-- NOTE[MM]: splitting a directory and taking last is a bit more involved
+--           than running Text.isInfixOf directly, but is more correct
+--           if e.g. username is LunaStudio - in that case, every path
+--           would be deleted
+createdByLunaStudio :: FilePath -> Bool
+createdByLunaStudio = (lunaStudio `Text.isInfixOf`) . Shelly.toTextIgnore . last . splitDirectories
```
[the remark this answers is not in the document]
Comment (mikusp) [reply luna-manager/src/Luna/Manager/Command/Uninstall.hs]: Good point, I'll change it even though I'm pretty sure splitDirectories won't return an empty list

File: luna-manager/src/Luna/Manager/Command/Options.hs [hunk 8]
```diff
@@ -96,11 +97,12 @@ evalOptionsParserT m = evalStateT m =<< parseOptions
 
 parseOptions :: MonadIO m => m Options
 parseOptions = liftIO $ customExecParser (prefs showHelpOnEmpty) optsParser where
-    commands           = mconcat [cmdInstall, cmdMkpkg, cmdUpdate, cmdDevelop, cmdSwitchVersion, cmdNextVer]
+    commands           = mconcat [cmdInstall, cmdUninstall, cmdMkpkg, cmdUpdate, cmdDevelop, cmdSwitchVersion, cmdNextVer]
     optsParser         = info (helper <*> optsProgram) (fullDesc <> header ("Luna ecosystem manager (" <> Info.version <> ")") <> progDesc Info.synopsis)
 
     -- Commands
     cmdInstall         = Opts.command "install"        . info optsInstall       $ progDesc "Install components. By default displays only the release versions."
+    cmdUninstall       = Opts.command "uninstall"      . info (pure Uninstall)  $ progDesc "Uninstall Luna Studio completely"
```
[the remark this answers is not in the document]
Comment (mikusp) [reply luna-manager/src/Luna/Manager/Command/Options.hs]: Definitely

**Pull request merged**